chalk-solve/src/clauses/program_clauses.rs

```diff
@@ -0,0 +1,871 @@
+use crate::RustIrDatabase;
+use chalk_ir::cast::{Cast, Caster};
+use chalk_ir::fold::shift::Shift;
+use chalk_ir::fold::Subst;
+use chalk_ir::*;
+use chalk_rust_ir::*;
+use std::iter;
+
+/// Trait for lowering a given piece of rust-ir source (e.g., an impl
+/// or struct definition) into its associated "program clauses" --
+/// that is, into the lowered, logical rules that it defines.
+pub trait ToProgramClauses {
```

Comment:
```suggestion
pub(super) trait ToProgramClauses {
```

doesn't matter, but it's nice to be precise

Reply:
`ToProgramClauses` is still used elsewhere, as noted in other comments.